luaui/Include/blueprint_substitution/logic.lua

```diff
@@ -0,0 +1,369 @@
+-- luaui/Include/blueprint_substitution/logic.lua
+-- Contains substitution logic and data, intended to be included by api_blueprint.lua widget
+local defPath = "luaui/Include/blueprint_substitution/definitions.lua"
+local BpDefs = VFS.Include(defPath)
+if not BpDefs then
+    error("Failed to load required blueprint definitions from: " .. defPath)
+end
+
+WG.BlueprintSubstitutionLogic = WG.BlueprintSubstitutionLogic or {}
+local BlueprintSubLogic = WG.BlueprintSubstitutionLogic -- Local alias for convenience
+
+local isDevModeParam = Spring.GetGameRulesParam('isDevMode')
+local ENABLE_REPORTS = (isDevModeParam == 1 or isDevModeParam == true or isDevModeParam == "1")
+local DEBUG_MODE = ENABLE_REPORTS
+
+BlueprintSubLogic.SIDES = BpDefs.SIDES
+
+BpDefs.defineUnitCategories()
+
+-- Load categories from the definitions file
+BlueprintSubLogic.UNIT_CATEGORIES = BpDefs.UNIT_CATEGORIES
+BlueprintSubLogic.categoryUnits = BpDefs.categoryUnits
+BlueprintSubLogic.unitCategories = BpDefs.unitCategories
+BlueprintSubLogic.equivalentUnits = {}
+BlueprintSubLogic.MasterBuildingData = {}
+BlueprintSubLogic.initialized = false
+
+-- Precompute map for faster lookups
+local unitNameToDefIDMap = {}
+Spring.Log("BlueprintSubLogic", "info", "Precomputing UnitDefs Name->ID map...")
+for defID, def in pairs(UnitDefs) do
+    if def and def.name then
+        unitNameToDefIDMap[def.name:lower()] = defID
+    end
+end
+Spring.Log("BlueprintSubLogic", "info", "Finished precomputing Name->ID map.")
+
```

Comment:
This was useful to "enforce" constants, but is only strict during debug, and even then it just errors louder/does more checks.

Reply:
What's the worst that can happen if you're not strict?

Reply:
Someone commits a file with multiple entries for the same category, where the first category is mistakenly ghosted and everyone ignores the log warnings/errors during debug.